File: sciencedirect.py
from articleparser import ArticleParser, ParserException
from bs4 import BeautifulSoup
import re


class ScienceDirectParser(ArticleParser):

    def get_doi(self):
        doi_link = self.soup.find('a', {'class': 'doi'}).text
        doi = '/'.join(doi_link.split('/')[-2:])
        return doi

    def have_access(self):
        dl_button = self.soup.find('span', {'class': 'pdf-download-label'})
        try:
            return dl_button.text == 'Download PDF'
        except AttributeError:
            return False

    def get_abstract(self):
        abstract = self.soup.find('div', {'class':
                                          'abstract author'}).find_all('p')
        if abstract is None:
            raise ParserException("Abstract not found")
        return ''.join(str(a) for a in abstract)

    def clean_html(self, soup):
        for s in [soup] + soup.find_all(True):
            class_attr = s.get('class')
            if class_attr == ['workspace-trigger']:
                if s['href'].startswith('#bib'):
                    rnum = s['href'][4:]
                    s.clear()
                    s.name = 'span'
                    s['data-refnum'] = rnum
                    s['class'] = 'ref'
                elif (s['href'].startswith('#fig') or
                      s['href'].startswith('#mmc')):
                    s['class'] = 'figure_ref'
                    s.name = 'span'
                    inner_html = s.string.encode('utf-8')
                    outer_html = s.encode('utf-8')
                    parent_html = s.parent.encode('utf-8')
                    s.clear()
                    fig_num = inner_html[::-1].find(b' ')
                    if fig_num == -1:
                        fig_num = inner_html[::-1].find(b'\xc2\xa0'[::-1])
                    if fig_num == -1:
                        fig_num = 0
                    fig_num = inner_html[-fig_num:]
                    fig_panel = []
                    fp_start = (parent_html.find(outer_html) +
                                len(outer_html))
                    fp_end_p = parent_html[fp_start:].find(b')')
                    fp_end_s = parent_html[fp_start:].find(b'<a')
                    fp_end_f = parent_html[fp_start:].find(b'.')
                    fp_end = -1
                    for end in [fp_end_p, fp_end_s, fp_end_f]:
                        if end != -1:
                            if fp_end == -1:
                                fp_end = end
                            else:
                                fp_end = min(end, fp_end)
                    print(fig_num)
                    print(parent_html[fp_start: fp_start + fp_end])
                    if fp_end == -1:
                        fp_end = 0
                        fig_panel = []
                        print(parent_html, fp_start)
                    for i in range(fp_start, fp_start + fp_end):
                        print(parent_html[i], parent_html[i-1])
                        if (parent_html[i] >= 65 and parent_html[i] <= 90):
                            if ((parent_html[i-1] >= 48 and
                                    parent_html[i-1] <= 57) or
                                    parent_html[i-1] == 62):
                                fig_panel.append(bytes([parent_html[i]]))
                            print(fig_panel)
                    fig_num = fig_num + b','.join(fig_panel)
                    s['data-fignum'] = fig_num.decode('utf-8')
                    if s['href'].startswith('#mmc'):
                        s['data-fignum'] = (s['data-fignum']
                                            + '-' + s['href'][1:])
                    if str(s).startswith('Table'):
                        s['data-fignum'] = 'T' + s['data-fignum']
                del s['href']
                del s['name']
            elif s.name in ['span', 'a']:
                s.replaceWithChildren()
            elif s.name in ['figure', 'p', 'div', 'tr', 'h2', 'h3']:
                del s['class']
                del s['id']
        html = str(soup)
        html = re.sub(r'<h[23]>(.*)</h[23]>', r'\1', html)
        html = re.sub(r'data-fignum="([\-mcA-Z1-9,]*)"></span>[^\)]*'
                      r'<span class="figure_ref"',
                      r'data-fignum="\1"></span><span class="figure_ref"',
                      html)
        html = re.sub(r'data-fignum="([\-mcA-Z1-9,]*)"></span>[^<]*\)',
                      r'data-fignum="\1"></span>)', html)
        html = re.sub(r' ?\(?<span', '<span', html)
        html = re.sub(r'</span>[;, ]', '</span>', html)
        html = re.sub(r'</span>\)', '</span>', html)
        html = re.sub(r'data-fignum="([\-mcA-Z1-9,]*)">'
                      r'</span>[^<^ ^\.^,]([ \.,])',
                      r'data-fignum="\1"></span>\2', html)
        return html

    def get_content(self):
        soup = None
        for s in self.soup.find_all('section'):
            if s.get('id') is not None and s['id'].startswith('ack'):
                soup = s
                break
        if soup is None:
            raise ParserException("Acknowledgments not found")
        ack = {'title': 'Acknowledgments',
                        'content': ''.join(self.clean_html(a) for a in
                                           soup.find_all('p'))}
        soup.clear()
        section = 0
        content = []
        while True:
            section += 1
            soup = self.soup.find('section', {'id': f'sec{str(section)}'})
            if soup is None:
                if section == 1:
                    raise ParserException("No content found")
                break
            soup = BeautifulSoup(soup.encode('utf-8'), 'lxml')
            if soup.h2 is not None:
                sec_title = self.clean_html(soup.h2)
            else:
                sec_title = None
            for elem in soup.find_all('figure'):
                elem.clear()
            content.append({'title': sec_title,
                            'content': []})
            subsect = 0
            while True:
                if sec_title is None:
                    break
                subsect += 1
                subsoup = soup.find('section',
                                    {'id':
                                     f'sec{str(section)}.{str(subsect)}'})
                if subsoup is None:
                    break
                subsect_title = self.clean_html(subsoup.h3)
                if subsect_title is None:
                    subsect_title = ''.join(self.clean_html(a) for a in
                                            subsoup.h3.contents)
                subsect_content = ''.join(self.clean_html(a) for a in
                                          subsoup.find_all(['p', 'figure']))
                content[-1]['content'].append({'title': subsect_title,
                                               'content': subsect_content})
            if content[-1]['content'] == []:
                content[-1]['content'] = ''.join(self.clean_html(a)
                                                 for a in soup.find_all(
                                                     ['p', 'figure']))
        content.append(ack)
        return content

    def get_references(self):
        soup = None
        for soup in self.soup.find_all('section'):
            css_class = soup.get('class')
            if isinstance(css_class, list) and css_class[0] == 'bibliography':
                break
        if soup is None:
            raise ParserException("References Not Found")
        labels = [a.text for a in soup.find_all('dt', {'class': 'label'})]
        ref_list = []
        count = 0
        for r in soup.find_all('dd'):
            if r['id'].find('sref') != -1:
                auth_title = r.find('div', {'class': 'contribution'})
                title = auth_title.strong.text
                auth_title.strong.clear()
                authors = auth_title.text.split(', ')
                journal_year = r.find('div', {'class': 'host'}).text
                journal = journal_year[:journal_year.find(',')]
                year = journal_year[journal_year.find('(') + 1:
                                    journal_year.find(')')]
                ref = {'label': labels[count],
                       'title': title,
                       'authors': authors,
                       'journal': journal,
                       'year': year}
                for a in r.find_all('a'):
                    doi = a['href'].find('doi.org/')
                    if doi != -1:
                        ref['doi'] = a['href'][doi + 8:]
                ref_list.append(ref)
                count += 1
        return ref_list

    def get_figures(self):
        figs = self.soup.find_all('figure')
        if len(figs) == 0:
            raise ParserException("Figures not found")
        fig_dict = {}
        for fig in figs:
            if fig['id'] == 'undfig1':
                name = 'Graphical Abstract'
            else:
                fignum = fig['id'][3:].upper()
                name = 'Figure ' + fignum
            fig_dict[name] = {}
            fig_text = fig.find_all('p')
            if len(fig_text) != 0:
                fig_dict[name]['title'] = self.clean_html(fig_text[0])
                fig_dict[name]['caption'] = ''.join(self.clean_html(a) for a in
                                                    fig_text[1:])
                if fig_dict[name]['title'].startswith('<p>'):
                    fig_dict[name]['title'] = fig_dict[name]['title'][3:-4]
            else:
                fig_dict[name]['title'] = 'Graphical Abstract'
                fig_dict[name]['caption'] = ''
            for link in fig.find_all('a'):
                if link.text.find('high-res') != -1:
                    fig_dict[name]['hr'] = link['href']
                elif link.text.find('full-size') != -1:
                    fig_dict[name]['lr'] = link['href']
        return fig_dict

    def get_files(self):
        soup = self.soup.find('div', {'class': 'PdfDownloadButton'})
        link = soup.find('a')
        if link is None:
            self.inject('script', 'document.getElementById("pdfLink").click()')
            raise ParserException("Clicked PDF download button")
        result = {}
        status, content, name = self.fetch(link['href'])
        soup = BeautifulSoup(content, 'lxml')
        pdf = soup.find_all('a')
        result['pdf'] = pdf[0]['href']
        soup = self.soup.find('div', {'class': 'Appendices'})
        links = []
        for a in soup.find_all('a'):
            if (a['href'] not in links and (a.get('title') is None or not
                                            a['title'].startswith('Help'))):
                links.append(a['href'])
        for i, a in enumerate(soup.find_all('span', {'class': 'label'})):
            result[a.text] = links[i]
        for key in result:
            if key.find('Supplemental Information') != -1:
                result['extended'] = result.pop(key)
                break
        return result


if __name__ == '__main__':
    s = ScienceDirectParser('https://www.sciencedirect.com/science/article/'
                            'abs/pii/S1097276519307294', debug=True)
    with open('output.html', 'w', encoding='utf-8') as f:
        for n, sect in s.article.content.items():
            f.write(f'<h2>{n}</h2>\n')
            if isinstance(sect, dict):
                for n2, subsect in sect.items():
                    f.write(f'<h3>{n2}</h3>\n')
                    f.write(subsect)
            else:
                f.write(sect)

File: assets/scrapers/sciencedirect.js
/*jshint esversion: 6 */

//
// Internal Functions
//

function ref_selector(elem) {
  var stuff = elem.querySelectorAll("a.workspace-trigger");
  var refs = [];
  for (var s of stuff) {
    if (s.name.startsWith("bb")) {
      refs.push(s);
    }
  }
  return refs;
}

function ref_matcher(text) {
  var m = text.match(
    /([,;]\s)?([^\(^\s]+(\set\sal\.|\sand\s[^\(^\s]+)?,\s\d\d\d\d)+/);
  if (m != null) {
    return m[0];
  } else {
    return null;
  }
}

function ref_num(ref, text) {
  return ref.name.slice(4);
}

function fig_ref_selector(elem) {
  var stuff = elem.querySelectorAll("a.workspace-trigger");
  var fig_refs = [];
  var is_figref = false;
  for (var s of stuff) {
    if (s.name.startsWith("bfig") || s.name.startsWith("bmmc") ||
        s.name.startsWith("bapp")) {
      if (s.textContent.startsWith("Fig")) {
        is_figref = true;
      } else if (s.textContent.startsWith("Table")) {
        is_figref = false;
      }
      if (is_figref) {
        fig_refs.push(s);
      }
    }
  }
  return fig_refs;
}

function fig_ref_matcher(text) {
  var m = text.match(
    /(Figures?|Figs?.|^)((,|;|,?\sand)?(\s|^)S?\d[A-Z]?(-S?\d?[A-Z]?)?)+/g);
  if (m != null) {
    return m[0];
  } else {
    return null;
  }
}

function fig_ref_num(ref, text) {
  var start = text.indexOf(ref.textContent);
  var num;
  if (start == -1) {
    return "";
  }
  num = text.slice(start);
  console.log(text);
  console.log(num);
  var m =  num.match(/((,|,?\sand)?\sS?\d+[A-Z]?)+/);
  if (m != null) {
    console.log(m[0]);
    m = m[0].replace(/\s/g, "");
    console.log(m);
    m = m.replace(/,?and/g, ",");
    console.log(m);
    var m_list = m.split(",");
    console.log(m_list);
    var prefix;
    if (m_list[0].length > 0) {
      prefix = m_list[0].match(/S?\d+/)[0];
    } else {
      prefix = m_list[1].match(/S?\d+/)[0];
    }
    for (var i = 0; i < m_list.length; i++) {
      if (!m_list[i].startsWith(prefix)) {
        m_list.splice(i, 1);
        i--;
      }
    }
    return m_list.join(",");
  } else {
    return "";
  }
}

function table_ref_selector(elem) {
  var stuff = elem.querySelectorAll("a.workspace-trigger");
  var table_refs = [];
  var is_table = false;
  for (var s of stuff) {
    if (s.name.startsWith("bfig") || s.name.startsWith("bmmc") ||
        s.name.startsWith("bapp")) {
      if (s.textContent.startsWith("Fig")) {
        is_table = false;
      } else if (s.textContent.startsWith("Table")) {
        is_table = true;
      }
      if (is_table) {
        table_refs.push(s);
      }
    }
  }
  return table_refs;
}

function table_ref_matcher(text) {
  var m = text.match(/(Tables?|^)((,|;|\s,?and)?(\s|^)S?\d[A-Z]?(-S?\d?[A-Z]?)?)+/g);
  if (m != null) {
    return m[0];
  } else {
    return null;
  }
}

var handlers = [{selector: ref_selector,
                 matcher: ref_matcher,
                 num: ref_num,
                 class: "ref"},
                {selector: fig_ref_selector,
                 matcher: fig_ref_matcher,
                 num: fig_ref_num,
                 class: "figure-ref"},
                 {selector: table_ref_selector,
                  matcher: table_ref_matcher,
                  num: fig_ref_num,
                  class: "table-ref"}];

//
// PyRead functions
//

function get_identifiers() {
  var doi_elem = document.querySelector("a.doi");
  var link = doi_elem.href;
  console.log(link);
  var slashes = [...link.matchAll(/\//g)];
  var doi_start = slashes[slashes.length - 2].index + 1;
  var doi = link.slice(doi_start);
  var title = document.querySelector("span.title-text").textContent;
  return {"doi": doi, "title": title};
}

function have_access() {
  var pdf_dl_button = document.querySelector("span.pdf-download-label");
  return pdf_dl_button.textContent == "Download PDF";
}

function get_abstract() {
  var abs = document.querySelector("div.abstract.author");
  if (abs == null) {
    abs = document.querySelector("div[id^=aep-abstract-sec]");
  } else {
    abs = abs.querySelector("div");
  }
  return abs;
}

function get_figures() {
  var figures = document.querySelectorAll("figure");
  var result = [];
  for (var i = 0; i < figures.length; i++) {
    var fig_entry = {};
    var fig_links = figures[i].querySelector("ol.links-for-figure");
    fig_links = fig_links.querySelectorAll("a");
    for (var j = 0; j < fig_links.length; j++) {
      if (fig_links[j].textContent.indexOf("high-res") == -1) {
        fig_entry.lr = fig_links[j].href;
      } else {
        fig_entry.hr = fig_links[j].href;
      }
    }
    var captions = figures[i].querySelector("span.captions");
    if (captions) {
      var caption_list = captions.querySelectorAll("p");
      if (caption_list.length > 1) {
        fig_entry.title = caption_list[0].textContent;
        caption_list[0].parentElement.removeChild(caption_list[0]);
      } else {
        var title = caption_list[0].textContent.match(
          /Fig\.\s\d+\.\s[^\.]*\.\s/);
        if (title != null) {
          fig_entry.title = title[0];
          captions.textContent = captions.textContent.slice(title[0].length);
        }
      }
      fig_entry.legend = handle_figs_refs(captions);
    } else {
      fig_entry.title = "Graphical Abstract";
    }
    result.push(fig_entry);
  }
  return result;
}

function get_content() {
  var content = [];
  var elems = document.querySelectorAll("section[id^=sec]");
  if (elems.length == 0) {
    var main = document.querySelector("div#body");
    console.log(main.textContent);
    if (main.textContent != "Loading...") {
      content = [{"title": "Body", "content": handle_figs_refs(main)}];
    }
  }
  for (var i = 0; i < elems.length; i++) {
    if (elems[i].id.indexOf(".") != -1 ||
        elems[i].parentElement.tagName == "SECTION") {
      //Skipping subsections (Which have '.' in their id)
      continue;
    }
    var section = {};
    var title = elems[i].querySelector("h2");
    if (title == null) {
      title = elems[i].querySelector("h3");
    }
    console.log(elems[i]);
    section.title = title.innerHTML;
    elems[i].removeChild(title);
    var subsects = elems[i].querySelectorAll("section");
    if (subsects.length == 0) {
      section.content = handle_figs_refs(elems[i]);
    } else {
      section.content = [];
      for (var j = 0; j < subsects.length; j++) {
        if ([...subsects[j].id.matchAll(/\./g)].length > 1) {
          continue;
        }
        var subsection = {};
        var subtitle = subsects[j].querySelector("h3");
        if (subtitle == null) {
          subtitle = subsects[j].querySelector("h4");
        }
        console.log(subsects[j]);
        subsection.title = subtitle.innerHTML;
        subsects[j].removeChild(subtitle);
        subsection.content = handle_figs_refs(subsects[j]);
        section.content.push(subsection);
      }
    }
    content.push(section);
  }
  return content;
}

function get_references() {
  var bib_section = document.querySelector("section[class^=bibliography]");
  var refs = bib_section.querySelectorAll("dt.label");
  var ref_list = [];
  for (var i = 0; i < refs.length; i++) {
    var refnum = refs[i].querySelector("a").href;
    refnum = parseInt(refnum.slice(refnum.indexOf('#bb') + 5), 10);
    var ref_entry = {};
    ref_entry.label = refs[i].textContent;
    var ref_info = refs[i].nextSibling;
    var auth_list = ref_info.querySelector("div.contribution");
    var journal_year, auth_list_end;
    if (auth_list) {
      auth_list = auth_list.textContent;
      ref_entry.title = ref_info.querySelector("strong");
      if (ref_entry.title != null) {
        ref_entry.title = ref_entry.title.textContent;
        auth_list_end = auth_list.indexOf(ref_entry.title);
        if (auth_list_end != -1) {
          auth_list = auth_list.slice(0, auth_list_end);
        }
      }
      journal_year = ref_info.querySelector("div.host").textContent;
    } else {
      var text = ref_info.textContent;
      auth_list_end = text.indexOf(" (");
      if (auth_list_end != -1) {
        auth_list = text.slice(0, auth_list_end);
      }
      var title_start = text.indexOf("). ") + 3;
      text = text.slice(title_start);
      title_end = text.indexOf(". ");
      ref_entry.title = text.slice(0, title_end);
    }
    if (journal_year != null) {
      var journal_end = journal_year.indexOf(",");
      var year_start = journal_year.slice(journal_end).indexOf('(') +
                       journal_end + 1;
      var year_end = journal_year.slice(year_start).indexOf(')') + year_start;
      ref_entry.journal = journal_year.slice(0, journal_end);
      ref_entry.year = journal_year.slice(year_start, year_end);
    }
    if (auth_list != null) {
      ref_entry.authors = auth_list.split(", ");
    }
    var links = ref_info.querySelectorAll("a");
    for (var j = 0; j < links.length; j++) {
      var doi_start = links[j].href.indexOf("doi.org/");
      if (doi_start != -1) {
        ref_entry.doi = links[j].href.slice(doi_start + 8);
        break;
      }
    }
    ref_list[refnum - 1] = ref_entry;
  }
  return ref_list;
}

function get_files() {
  result = {};
  var pdf_button = document.querySelector("div.PdfDownloadButton");
  var link = pdf_button.querySelector("a");
  if(!link) {
    document.getElementById("pdfLink").click();
    return {};
  } else {
    result.pdf = link.href;
    if (result.pdf != null) {
      var xhr = new XMLHttpRequest();
      xhr.open("GET", result.pdf, false);
      xhr.send();
      var response = xhr.responseText;
      var start = response.indexOf('<a href="') + 9;
      var end = response.slice(start).indexOf('">') + start;
      result.pdf = response.slice(start, end);
    }
    var appendix = document.querySelector("div.Appendices");
    if (appendix != null) {
      var links = appendix.querySelectorAll("span.article-attachment");
      for (var i = 0; i < links.length; i++) {
        if (links[i].nextSibling == null) {
          continue;
        }
        var title = links[i].nextSibling.textContent;
        if (title.indexOf("Supplemental Information") != -1 &&
            !result.hasOwnProperty("extended")) {
          title = "extended";
        }
        link = links[i].querySelector("a");
        result[title] = link.href;
      }
    }
  }
  return result;
}
